docs/rfc/0000-application-architecture.md

```diff
@@ -0,0 +1,503 @@
+
+Each subcommand MUST expose its functionality as a linkable library, and provide
+CBOR serialisation for its arguments and outputs.
+
+To enable [browser applications](#browser-applications), subcommands may be
+callable over the [IPC](#ipc) protocol ("CLIaaS"). The canonical CLI may provide
+a command to bind all available subcommands to a socket wholesale, or a
+configurable subset of them. Unprivileged subcommands may also be exposed as a
+HTTP API (see [Browser Applications](#browser-applications)).
+
+The builtin subcommands shall include network clients for interacting with local
+p2p nodes and the pub-sub bus. The vast majority of new functionality will,
+however, be built on top of the [`Storage`](#storage) primitive. If and when it
+becomes desirable to prototype such functionality in other programming languages
+(ie. without linking to the storage library directly) it will be considered to
+expose the storage interface as a network service itself.
+
+> Note that it does not make much sense to proxy networked subcommands when
+> binding to a socket. Frontends will need to connect to the respective services
+> directly.
+
+A comprehensive specification of the (initial) set of builtins is beyond the
+scope of this document.
+
+
+### Browser Applications
+
+In practice, browser applications are limited to HTTP or WebSockets for network
+calls, filesystem operations must be delegated to a backend service. This poses
+two security concerns:
+
+1. The backend must bind to a well-known TCP port, which is then accessible by
+   all other programs on the same host. Moreover, browsers can be tricked by
+   remote sites into issuing requests against the local host.
+2. Public-key based authentication APIs (such as [WebAuthn][webauthn]) are not
+   yet mature / flexible / portable enough to allow a browser frontend to prove
+   access to the (shared) private key to a backend process.
+
+For these reasons, we devise the following restrictions:
+
+* TCP-based backend services MUST NOT expose privileged operations.
+* Applications built using browser-based frameworks (such as [electron]) MUST
+  proxy service calls through the privileged `main` process, which is
+  responsible for acquiring signatures via the `ssh-agent` protocol where
+  necessary, and calling the backend service via [IPC](#ipc).
+
+Since preferences of web developers regarding the granularity of remote calls
+vary, they may choose to implement their own IPC backend, possibly re-composing
+the library functionality of [CLI](#cli) subcommands.
+
+Browser applications SHOULD be packaged in a way such that their main backend
+conforms to [Process Orchestration](#process-orchestration), and include other
+services via the package manager's dependency management mechanism. If that is
+not desired (e.g. [AppImage] bundles), they MUST sandbox all service
+dependencies, and MUST NOT interact with services installed outside the sandbox
```

Comment:
AppImage’s philosophy is decidedly to bundle everything in one image, and _not_ provide any dependency management. Sure you can stick an AppImage into a distribution package, but I don’t think this would be accepted in most distros.

So, if that’s the case, then the AppImage cannot assume anything about the environment it is running in — and consequently should also not modify it in any way (eg by installing things outside of its sandbox). 

Of course one can say “install packages x, y, z using your system package manager before being able to use this”, but this seems to defeat the whole idea of self-containedness.

Reply:
Perhaps the main problem with the current architecture is that upstream manages its own p2p stack — there should only be one instance per key, so if there are other apps on the same machine, they either need to talk to upstream, or use their own key.

Offline operations are less problematic — but then, if your AppImage _assumes_ the networking service to be present, yet writes to the storage directly, I’d expect this to result in a mess to install and configure.

Reply:
Here’s another approach: say we use lock files to prevent multiple instances of the network stack. Being a good app bundle, the app bundles all executables it needs, implements its own process management etc. Being a good citizen in the radicle app ecosystem, its interfaces to the outside world are the `ssh-agent` and DBus (because it wants to know about pushes). 

This would mean DBus is a hard requirement. It would also mean that we can’t use DBus to activate services, because which one will respond? If we decide that DBus is not the right fit, because it’s too slow, not portable enough or whatever, we’re back to square one because that component must be made available to all radicle apps.

So idk, but self-contained app bundles just don’t seem to be a good fit if we’re aiming for a diverse ecosystem (unless they are truly self-contained that is).

Reply:
> Sure you can stick an AppImage into a distribution package, but I don’t think this would be accepted in most distros.

This was exactly what I was getting at. I wanted to know whether there are any reasons not to use AppImage even if we distribute it via package managers and don’t support it standalone. The reasoning behind this is that I wasn’t sure how easy it is to bundle electron apps without relying on AppImage. (Spoiler: it’s easy)

In general I agree that we should distribute applications so that they integrate naturally with other Radicle services. The best way to accomplish that is indeed package managers and service managers.

Reply:
I dropped this. Perhaps we should have a separate proposal for packaging guidelines, maybe in `radicle-decisions` instead?

Reply:
Sounds good to me :+1: